lib/phoenix_live_view/channel.ex

```diff
@@ -157,41 +213,138 @@ defmodule Phoenix.LiveView.Channel do
+      :diff ->
+        {diff, new_state} = render_diff(new_state, new_socket)
+
+        {:ok, :diff,
+         new_state
+         |> push_internal_live_redirect(pending_internal_live_redirect, nil)
+         |> push_render(diff, ref)}
+
+      :noop ->
+        {:ok, :noop,
+         state
+         |> push_internal_live_redirect(pending_internal_live_redirect, nil)
+         |> push_noop(ref)}
+
+      {:redirect, %{to: to} = opts} ->
+        send(new_state.transport_pid, {:socket_close, self(), {:redirect, to}})
+        {:stop, {:shutdown, {:redirect, to}}, push_redirect(new_state, opts, ref)}
+
+      {:live_redirect, {:internal, params}, %{to: _to, kind: _kind} = opts} ->
+        state
+        |> drop_redirect()
+        |> sync_handle_params_with_live_redirect(params, opts, ref)
+
+      {:live_redirect, :external, %{to: to} = opts} ->
+        send(state.transport_pid, {:socket_close, self(), {:redirect, to}})
+        opts = Map.put(opts, :flash, View.sign_flash(state.socket, View.get_flash(state.socket)))
+        {:stop, {:shutdown, {:redirect, to}}, push_external_live_redirect(state, opts, ref)}
     end
   end
 
-  defp push_render(state, socket) do
-    case maybe_diff(state, socket) do
-      {diff, new_state} -> push(new_state, "render", diff)
-      :noop -> state
+  defp drop_redirect(state) do
+    %{state | socket: View.drop_redirect(state.socket)}
+  end
+
+  defp sync_handle_params_with_live_redirect(state, params, opts, ref) do
+    if function_exported?(state.socket.view, :handle_params, 2) do
+      case state.socket.view.handle_params(params, state.socket) do
+        {:noreply, %Socket{} = new_socket} ->
+          handle_changed(state, new_socket, ref, opts)
+
+        {:stop, %Socket{} = new_socket} ->
+          case handle_changed(state, new_socket, ref, opts) do
+            {:ok, _changed, new_state} -> {:stop, :shutdown, new_state}
+            {:stop, reason, new_state} -> {:stop, reason, new_state}
+          end
+      end
+    else
+      {:ok, :live_redirect, push_internal_live_redirect(state, opts, ref)}
     end
   end
 
-  defp push_redirect(%{socket: socket} = state, to, flash) do
-    push(state, "redirect", %{to: to, flash: View.sign_flash(socket, flash)})
+  defp push_internal_live_redirect(state, nil, _ref), do: state
+
+  defp push_internal_live_redirect(state, %{to: to, kind: kind}, nil = _ref) do
+    push(state, "live_redirect", %{to: to, kind: kind})
+  end
+
+  defp push_internal_live_redirect(state, %{to: to, kind: kind}, ref) do
+    reply(state, ref, :ok, %{live_redirect: %{to: to, kind: kind}})
+  end
+
+  defp push_redirect(state, %{to: to}, nil = _ref) do
+    flash = View.get_flash(state.socket)
+    push(state, "redirect", %{to: to, flash: View.sign_flash(state.socket, flash)})
+  end
+
+  defp push_redirect(state, %{to: to}, ref) do
+    flash = View.get_flash(state.socket)
+    reply(state, ref, :ok, %{to: to, flash: View.sign_flash(state.socket, flash)})
```

Comment:
Bug here? Don't you need to nest the 4th argument with :redirect? Maybe have a `push_or_reply` function.